Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp

```diff
@@ -1245,12 +1251,24 @@ webm::Status WebMParser::AudioTrackData::consumeFrameData(webm::Reader& reader,
         }
     }
 
+    ASSERT(formatDescription());
     if (!m_processedMediaSamples.info())
         m_processedMediaSamples.setInfo(formatDescription());
-    else if (formatDescription() && *formatDescription() != *m_processedMediaSamples.info())
+    else if (*formatDescription() != *m_processedMediaSamples.info())
         drainPendingSamples();
 
-    m_processedMediaSamples.append({ presentationTime, MediaTime::invalidTime(), m_packetDuration, WTFMove(m_completeFrameData), MediaSample::SampleFlags::IsSync });
+    MediaTime lastEndTime = (m_lastFrameTime.isValid() ? m_lastFrameTime : presentationTime) + MediaTime(m_frames, downcast<const AudioInfo>(*formatDescription()).rate);
+    // If the gap between the real previous frame's end and the new presentation time is less than the smallest possible value in the presentation's timescale
+    // causing the frames to overlap, we can assume a rounding error has occurred in the container and we will calculate the actual presentation time
+    // according to the number of frames parsed so far.
+    if (!m_lastFrameTime.isValid() || (m_lastFrameTime != presentationTime && (presentationTime.timeScale() == lastEndTime.timeScale() || presentationTime >= lastEndTime || lastEndTime - presentationTime >= MediaTime(1, presentationTime.timeScale())))) {
+        // We are starting a new block
+        m_frames = 0;
+        lastEndTime = m_lastFrameTime = presentationTime;
```

Comment:
After doing some local testing, whenever we hit this logic, our timescale changes from the audio sample rate (441000 for example) to the video timescale (1000 in most cases).

Here is a sample output log:
`start = lastEndTime`
`end = lastEndTime + m_packetDuration`
```
2022-08-18 12:17:45.474578-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.65663945578231,"numerator":20575578,"denominator":441000,"flags":1}
2022-08-18 12:17:45.474660-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.679859410430836,"numerator":20585818,"denominator":441000,"flags":1}
2022-08-18 12:17:45.474751-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.669,"numerator":46669,"denominator":1000,"flags":1}
2022-08-18 12:17:45.474830-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.69221995464853,"numerator":20591269,"denominator":441000,"flags":1}
2022-08-18 12:17:45.474971-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.672,"numerator":46672,"denominator":1000,"flags":1}
2022-08-18 12:17:45.475065-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.69521995464853,"numerator":20592592,"denominator":441000,"flags":1}
2022-08-18 12:17:45.475245-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.685,"numerator":46685,"denominator":1000,"flags":1}
2022-08-18 12:17:45.475326-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.70821995464853,"numerator":20598325,"denominator":441000,"flags":1}
2022-08-18 12:17:45.475456-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.698,"numerator":46698,"denominator":1000,"flags":1}
2022-08-18 12:17:45.475541-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.721219954648525,"numerator":20604058,"denominator":441000,"flags":1}
2022-08-18 12:17:45.475681-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.701,"numerator":46701,"denominator":1000,"flags":1}
2022-08-18 12:17:45.475769-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.724219954648525,"numerator":20605381,"denominator":441000,"flags":1}
2022-08-18 12:17:45.475907-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.704,"numerator":46704,"denominator":1000,"flags":1}
2022-08-18 12:17:45.475995-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.727219954648525,"numerator":20606704,"denominator":441000,"flags":1}
2022-08-18 12:17:45.476122-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.707,"numerator":46707,"denominator":1000,"flags":1}
2022-08-18 12:17:45.476205-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.730219954648526,"numerator":20608027,"denominator":441000,"flags":1}
2022-08-18 12:17:45.476335-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.72,"numerator":46720,"denominator":1000,"flags":1}
2022-08-18 12:17:45.476420-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.743219954648524,"numerator":20613760,"denominator":441000,"flags":1}
2022-08-18 12:17:45.476595-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.743219954648524,"numerator":20613760,"denominator":441000,"flags":1}
2022-08-18 12:17:45.476676-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.766439909297056,"numerator":20624000,"denominator":441000,"flags":1}
2022-08-18 12:17:45.476830-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.766439909297056,"numerator":20624000,"denominator":441000,"flags":1}
2022-08-18 12:17:45.476907-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.78965986394558,"numerator":20634240,"denominator":441000,"flags":1}
2022-08-18 12:17:45.477010-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.79,"numerator":46790,"denominator":1000,"flags":1}
2022-08-18 12:17:45.477089-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.813219954648524,"numerator":20644630,"denominator":441000,"flags":1}
2022-08-18 12:17:45.477275-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.813219954648524,"numerator":20644630,"denominator":441000,"flags":1}
2022-08-18 12:17:45.477354-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.83643990929705,"numerator":20654870,"denominator":441000,"flags":1}
2022-08-18 12:17:45.477459-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.83643990929705,"numerator":20654870,"denominator":441000,"flags":1}
2022-08-18 12:17:45.477535-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.85965986394558,"numerator":20665110,"denominator":441000,"flags":1}
2022-08-18 12:17:45.477683-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.85965986394558,"numerator":20665110,"denominator":441000,"flags":1}
2022-08-18 12:17:45.477768-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.882879818594105,"numerator":20675350,"denominator":441000,"flags":1}
2022-08-18 12:17:45.477863-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.883,"numerator":46883,"denominator":1000,"flags":1}
2022-08-18 12:17:45.477945-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.90621995464853,"numerator":20685643,"denominator":441000,"flags":1}
2022-08-18 12:17:45.478152-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD start = {"value":46.90621995464853,"numerator":20685643,"denominator":441000,"flags":1}
2022-08-18 12:17:45.478321-0600 com.apple.WebKit.GPU.Development[36126:3532751] J$*FD end = {"value":46.92943990929705,"numerator":20695883,"denominator":441000,"flags":1}
```
Every time we hit a new block, the start time switches timescale while the end time (which is calculated based off the duration with 441000 timescale) remains. So every once in a while you will hear the clipping due to the rounding errors.

Reply:
It also looks like we are assuming a new block is reached more often than you would expect (I think there may be an issue in the conditional statement).